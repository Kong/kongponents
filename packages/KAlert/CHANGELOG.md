# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.7](https://github.com/Kong/kongponents/compare/@kongponents/kalert@0.1.6...@kongponents/kalert@0.1.7) (2019-11-11)

**Note:** Version bump only for package @kongponents/kalert





## [0.1.6](https://github.com/Kong/kongponents/compare/@kongponents/kalert@0.1.5...@kongponents/kalert@0.1.6) (2019-11-08)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))





## [0.1.5](https://github.com/Kong/kongponents/compare/@kongponents/kalert@0.1.4...@kongponents/kalert@0.1.5) (2019-11-08)

**Note:** Version bump only for package @kongponents/kalert





# @kongponents/kalert

## Table of Contents

- [0.1.4](#134)

### 0.1.4
 - Updates close icon to use new close KIcon [#134](https://github.com/Kong/kongponents/pull/134)
