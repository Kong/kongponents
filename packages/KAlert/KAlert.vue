<template>
  <div
    v-if="isShowing"
    :class="[appearance, size, {'isBordered':isBordered}, {'hasLeftBorder':hasLeftBorder}, {'isCentered': isCentered}, {'isFixed': isFixed}]"
    class="k-alert"
    role="alert">
    <button
      v-if="isDismissible"
      type="button"
      aria-label="Close"
      class="close"
      @click="dismissAlert()">
      <svg
        xmlns="http://www.w3.org/2000/svg"
        width="12"
        height="12">
        <path
          fill="#000"
          fill-opacity=".25"
          fill-rule="evenodd"
          d="M0 1l1-1h.5L6 4.499939 10.5 0h.5l1 1v.5L7.5 6l4.5 4.5v.5l-1 1h-.5L6 7.5 1.5 12H1l-1-1v-.5L4.5 6 0 1.5z"/>
      </svg>
    </button>
    <span
      v-if="hasIcon"
      class="alert-icon">
      <slot name="alertIcon" />
    </span>
    <span>
      <slot name="alertMessage">{{ alertMessage }}</slot>
    </span>
  </div>
</template>

<script>
export const appearances = {
  info: 'info',
  success: 'success',
  danger: 'danger',
  warning: 'warning'
}

export default {
  name: 'KAlert',
  props: {
    /**
    * Message to show in alert
    */
    alertMessage: {
      type: String,
      default: ''
    },
    /**
     * Set if close button is visible
     */
    isDismissible: {
      type: Boolean,
      default: false
    },
    /**
     * Set whether or not the alert box is shown.
     */
    isShowing: {
      type: Boolean,
      default: true
    },
    /**
     * Fixes alert to top
     */
    isFixed: {
      type: Boolean,
      default: false
    },
    /**
     * Set whether or not alert has full border is visible
     */
    isBordered: {
      type: Boolean,
      default: false
    },
    /**
     * Sets whether or not alert has left border
     */
    hasLeftBorder: {
      type: Boolean,
      default: false
    },
    /**
     * Center text inside alert
     */
    isCentered: {
      type: Boolean,
      default: false
    },
    /**
      * Base styling of the button<br>
      * One of ['primary, danger, warning, success ]
      */
    appearance: {
      type: String,
      default: 'info',
      validator: function (value) {
        return Object.values(appearances).indexOf(value) !== -1
      }
    },
    /**
     * Set whether alert box is the default size or small for context (under form fields, etc),
     */
    size: {
      type: String,
      default: '',
      validator: (value) => {
        return [
          '',
          'small'
        ].indexOf(value) !== -1
      }
    }
  },

  computed: {
    hasIcon () {
      return !!this.$slots.alertIcon
    }
  },

  methods: {
    dismissAlert () {
      this.$emit('closed')
    }
  }
}
</script>

<style scoped lang="scss">
@import '~@kongponents/styles/_variables.scss';

.k-alert {
  position: relative;
  display: flex;
  align-items: center;
  padding: var(--spacing-sm, spacing(sm)) var(--spacing-md, spacing(md));
  font-family: inherit;
  font-size: 1rem;
  border-radius: 3px;
  a {
    text-decoration: underline;
    color: var(--blue-link, color(blue-link));
  }

  .alert-icon {
    margin-right: var(--spacing-xs, spacing(xs));
    max-height: 1rem;
    svg,
    img {
      max-height: 1rem;
    }
  }

  .close {
    position: absolute;
    top: 0;
    right: 18px;
    bottom: 0;
    border: 0;
    background-color: transparent;
    transition: all 200ms ease;
    cursor: pointer;
    &:hover,
    &:active {
      text-decoration: none;
      opacity: .75;
    }
  }

  // Variants
  &.isFixed {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }
  &.isBordered {
    border: 1px solid;
  }
  &.isCentered {
    justify-content: center;
  }
  &.hasLeftBorder {
   border-left: 3px solid;
   border-radius: 0;
  }
  &.small {
    font-size: var(--type-sm, type(sm));
    padding: var(--spacing-sm, spacing(sm)) var(--spacing-xs, spacing(xs));
  }

  // Appearances
  &.info {
    color: var(--KAlertInfoColor, var(--blue-dark, color(blue-dark)));
    border-color: var(--KAlertInfoBorder, var(--blue-light-01, color(blue-light-01)));
    background-color: var(--KAlertInfoBackground, var(--blue-lighter, color(blue-lighter)));
  }
  &.success {
    color: var(--KAlertSuccessColor, var(--green-dark, color(green-dark)));
    border-color: var(--KAlertSuccessBorder, var(--green-light-01, color(green-light-01)));
    background-color: var(--KAlertSuccessBackground, var(--green-lighter, color(green-lighter)));
  }
  &.danger {
    color: var(--KAlertDangerColor, var(--red-dark, color(red-dark)));
    border-color: var(--KAlertDangerBorder, var(--red-light-01, color(red-light-01)));
    background-color: var(--KAlertDangerBackground, var(--red-lighter, color(red-lighter)));
  }
  &.warning {
    color: var(--KAlertWarningColor, var(--yellow-darker, color(yellow-dark)));
    border-color: var(--KAlertWarningBorder, var(--yellow-light, color(yellow-light)));
    background-color: var(--KAlertWarningBackground, var(--yellow-lighter, color(yellow-lighter)));
  }
}
</style>
