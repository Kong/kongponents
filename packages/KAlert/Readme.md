```js
const attributes = {
  id: 'someIDName',
}
<KAlert isShowing alertMessage="howdy!" isDismissible :alertAttributes=attributes />
```
