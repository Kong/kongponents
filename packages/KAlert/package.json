{
  "name": "@kongponents/kalert",
  "version": "0.1.3",
  "description": "Alert component",
  "main": "KAlert.vue",
  "scripts": {
    "test": "echo \"Error: no test specified\""
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme"
}
