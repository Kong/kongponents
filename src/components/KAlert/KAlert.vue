<template>
  <div
    class="k-alert"
    :class="[appearance]"
  >
    <div
      v-if="!hideIcon"
      class="alert-icon-container"
    >
      <slot name="icon">
        <component
          :is="getAlertIcon"
          class="alert-icon"
          :size="KUI_ICON_SIZE_40"
        />
      </slot>
    </div>
    <div class="alert-content">
      <span
        v-if="title"
        class="alert-title"
      >
        {{ title }}
      </span>
      <div
        v-if="message || $slots.default"
        class="alert-message"
      >
        <slot>
          <p>{{ message }}</p>
        </slot>
      </div>
    </div>
    <CloseIcon
      v-if="dismissible"
      class="alert-dismiss-icon"
      :size="KUI_ICON_SIZE_40"
      tabindex="0"
      @click="$emit('dismiss')"
      @keydown.space.prevent
      @keyup.enter="$emit('dismiss')"
      @keyup.space="$emit('dismiss')"
    />
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue'
import type { PropType } from 'vue'
import type { AlertAppearance } from '@/types'
import { AlertAppearances } from '@/types'
import { InfoIcon, CheckCircleIcon, WarningIcon, ClearIcon, CloseIcon } from '@kong/icons'
import { KUI_ICON_SIZE_40 } from '@kong/design-tokens'

type AlertIcon = typeof InfoIcon | typeof CheckCircleIcon | typeof WarningIcon | typeof ClearIcon

const props = defineProps({
  title: {
    type: String,
    default: '',
  },
  message: {
    type: String,
    default: '',
  },
  appearance: {
    type: String as PropType<AlertAppearance>,
    default: 'info',
    validator: (value: AlertAppearance): boolean => {
      return Object.values(AlertAppearances).includes(value)
    },
  },
  hideIcon: {
    type: Boolean,
    default: true,
  },
  dismissible: {
    type: Boolean,
    default: false,
  },
})

defineEmits(['dismiss'])

const getAlertIcon = computed((): AlertIcon => {
  switch (props.appearance) {
    case AlertAppearances.info:
      return InfoIcon
    case AlertAppearances.success:
      return CheckCircleIcon
    case AlertAppearances.warning:
      return WarningIcon
    case AlertAppearances.danger:
      return ClearIcon
    default:
      return InfoIcon // info as default in case of invalid appearance
  }
})
</script>

<style lang="scss" scoped>
/* Component mixins */

@mixin kAlertAppearance($backgroundColor: var(--kui-color-background-primary-weakest, $kui-color-background-primary-weakest), $textColor: var(--kui-color-text-primary, $kui-color-text-primary), $dismissIconHoverColor: var(--kui-color-text-primary-strong, $kui-color-text-primary-strong)) {
  background-color: $backgroundColor;
  color: $textColor;

  .alert-dismiss-icon {
    &:hover,
    &:focus {
      color: $dismissIconHoverColor !important;
    }
  }
}

/* Component styles */
.k-alert {
  @include kAlertAppearance;

  border-radius: var(--kui-border-radius-30, $kui-border-radius-30);
  display: flex;
  gap: var(--kui-space-30, $kui-space-30);
  padding: var(--kui-space-50, $kui-space-50);

  .alert-icon-container {
    :deep(#{$kongponentsKongIconSelector}) { // target slotted icon
      height: var(--kui-icon-size-40, $kui-icon-size-40) !important;
      width: var(--kui-icon-size-40, $kui-icon-size-40) !important;
    }
  }

  .alert-content {
    align-self: center;
    display: flex;
    flex: 1;
    flex-direction: column;
    font-family: var(--kui-font-family-text, $kui-font-family-text);
    gap: var(--kui-space-30, $kui-space-30);

    .alert-title {
      font-size: var(--kui-font-size-50, $kui-font-size-50);
      font-weight: var(--kui-font-weight-semibold, $kui-font-weight-semibold);
      letter-spacing: var(--kui-letter-spacing-minus-30, $kui-letter-spacing-minus-30);
      line-height: var(--kui-line-height-40, $kui-line-height-40);
    }

    .alert-message {
      font-size: var(--kui-font-size-30, $kui-font-size-30);
      font-weight: var(--kui-font-weight-regular, $kui-font-weight-regular);
      line-height: var(--kui-line-height-30, $kui-line-height-30);

      p {
        margin: 0;
      }
    }
  }

  .alert-dismiss-icon {
    border-radius: var(--kui-border-radius-20, $kui-border-radius-20);
    cursor: pointer;
    margin-left: var(--kui-space-auto, $kui-space-auto);
    outline: none;

    &:focus-visible {
      box-shadow: var(--kui-shadow-focus, $kui-shadow-focus);
    }
  }

  &.info {
    @include kAlertAppearance;
  }

  &.success {
    @include kAlertAppearance(var(--kui-color-background-success-weakest, $kui-color-background-success-weakest), var(--kui-color-text-success, $kui-color-text-success), var(--kui-color-text-success-strong, $kui-color-text-success-strong));
  }

  &.warning {
    @include kAlertAppearance(var(--kui-color-background-warning-weakest, $kui-color-background-warning-weakest), var(--kui-color-text-warning, $kui-color-text-warning), var(--kui-color-text-warning-strong, $kui-color-text-warning-strong));
  }

  &.danger {
    @include kAlertAppearance(var(--kui-color-background-danger-weakest, $kui-color-background-danger-weakest), var(--kui-color-text-danger, $kui-color-text-danger), var(--kui-color-text-danger-strong, $kui-color-text-danger-strong));
  }
}
</style>
