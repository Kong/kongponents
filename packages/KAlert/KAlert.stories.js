import { storiesOf } from '../../storybook/node_modules/@storybook/vue'

import { default as KAlert, appearances } from './KAlert.vue'
import KIcon from '../KIcon/KIcon.vue'

storiesOf('KAlert', module)
  .add('Alert - Dismissible', () => ({
    components: { KAlert },
    template: `<KAlert :isShowing="isShowing" isDismissible alertMessage="howdy!" @closed="dismissAlert" />`,
    data: () => {
      return {
        isShowing: true
      }
    },
    methods: {
      dismissAlert () {
        console.log('dismissed!')
        this.isShowing = false
      }
    }
  }), {info: {}})
  .add('Alert - Centered', () => ({
    components: { KAlert },
    template: `<KAlert isCentered alertMessage="I'm centered!" />`
  }), {info: {}})
  .add('Alert - Fixed', () => ({
    components: { KAlert },
    template: `<div style="height:200vh">
      <KAlert isFixed alertMessage="I'm fixed to the top!" />
    </div>`
  }), {info: {}})
  .add('Alert - Appearances', () => ({
    components: { KAlert },
    template: `
    <div>
      ${Object.keys(appearances).map(type => {
      return `<KAlert isDismissible alertMessage="${type}" appearance="${type}" style="margin-bottom: 10px;"/>`
    }).join('')}
    </div>`
  }), {info: {}})
  .add('Alert - Bordered', () => ({
    components: { KAlert },
    template: `
    <div>
      ${Object.keys(appearances).map(type => {
      return `<KAlert isBordered alertMessage="${type}" appearance="${type}" style="margin-bottom: 10px;"/>`
    }).join('')}
    </div>`
  }), {info: {}})
  .add('Alert - Small & Left Border', () => ({
    components: { KAlert },
    template: `
    <div>
      ${Object.keys(appearances).map(type => {
      return `<div style="width:300px; margin-bottom: 10px;">
        <KAlert hasLeftBorder size="small" alertMessage="${type}" appearance="${type}"/>
      </div>`
    }).join('')}
    </div>`
  }), {info: {}})
  .add('Alert - Slots', () => ({
    components: { KAlert, KIcon },
    template: `
    <div>
      ${Object.keys(appearances).map(type => {
      return `<div style="margin-bottom: 10px;">
          <KAlert appearance="${type}">
            <template slot="alertIcon">
              <KIcon icon="portal" />
            </template>
            <template slot="alertMessage">
              I have an icon and a <a href="">Link</a>!
            </template>
          </KAlert>
        </div>`
    }).join('')}

    </div>`
  }), {info: {}})
