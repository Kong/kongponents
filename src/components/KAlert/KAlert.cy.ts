import { h } from 'vue'
import { mount } from 'cypress/vue'
import KAlert from '@/components/KAlert/KAlert.vue'
import { AlertAppearances } from '@/types'

const rendersCorrectVariant = (variant: string) => {
  it(`renders ${variant} variant`, () => {
    mount(KAlert, {
      props: {
        appearance: `${variant}`,
        alertMessage: `I am ${variant}`,
      },
    })

    cy.get('.k-alert').should('have.class', variant)
  })
}

describe('KAlert', () => {
  // Loop through AlertAppearances
  Object.keys(AlertAppearances).map(v => rendersCorrectVariant(v))

  it('renders info variant when no appearance prop', () => {
    mount(KAlert, {
      props: {
        alertMessage: 'I should be info!',
      },
    })

    cy.get('.k-alert').should('have.class', 'info')
  })

  it('renders all elements correctly when props are not passed', () => {
    mount(KAlert)

    cy.get('.k-alert').should('exist')
    cy.get('.alert-icon-container').should('not.exist')
    cy.get('.alert-content').should('exist')
    cy.get('.alert-title').should('not.exist')
    cy.get('.alert-message').should('not.exist')
    cy.get('.alert-dismiss-icon').should('not.exist')
  })

  it('renders title when passed as a prop', () => {
    const title = 'I am a title'

    mount(KAlert, {
      props: {
        title,
      },
    })

    cy.get('.alert-title').should('be.visible').and('have.text', title)
  })

  it('renders message when passed as a prop', () => {
    const message = 'I am a message'

    mount(KAlert, {
      props: {
        message,
      },
    })

    cy.get('.alert-message').should('be.visible').and('have.text', message)
  })

  it('renders icon and dismiss button when props are passed', () => {
    mount(KAlert, {
      props: {
        showIcon: true,
        dismissible: true,
      },
    })

    cy.get('.alert-icon-container').should('exist')
    cy.get('.alert-dismiss-icon').should('exist')
  })

  it('renders default slot correctly', () => {
    const message = 'I am a message'
    const defaultSlotContent = 'Default'

    mount(KAlert, {
      props: {
        message,
      },
      slots: {
        default: h('span', {}, defaultSlotContent),
      },
    })

    cy.get('.alert-message').should('be.visible').and('have.text', defaultSlotContent)
  })

  it('displays icon passed through icon slot', () => {
    const iconSlotContent = 'icon slot content'
    const testId = 'slotted-icon'

    mount(KAlert, {
      slots: {
        icon: h('div', { 'data-testid': testId }, iconSlotContent),
      },
    })

    cy.get('.alert-icon-container').findTestId(testId).should('be.visible')
    cy.get('.alert-icon-container').findTestId(testId).should('contain.text', iconSlotContent)
  })

  it('emits dismiss event when dismiss button is clicked', () => {
    mount(KAlert, {
      props: {
        dismissible: true,
      },
    })

    cy.get('.alert-dismiss-icon').click().then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'dismiss')
    })
  })
})
