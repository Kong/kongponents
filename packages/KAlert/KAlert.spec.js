import { mount } from '@vue/test-utils'
import { default as KAlert, appearances } from './KAlert.vue'

let rendersCorrectVariant = (variant) => {
  it(`Renders ${variant} variant`, () => {
    const wrapper = mount(KAlert, {
      propsData: {
        'appearance': `${variant}`,
        'message': `I am ${variant}`
      }
    })

    expect(wrapper.find('.k-alert').classes()).toContain(`${variant}`)
    expect(wrapper.html()).toMatchSnapshot()
  })
}

describe('KAlert', () => {
  Object.keys(appearances).map(v => rendersCorrectVariant(v))

  it('renders info variant when no appearance prop', () => {
    const wrapper = mount(KAlert, {
      propsData: {
        'message': 'I should be info!'
      }
    })

    expect(wrapper.find('.k-alert').classes()).toContain('info')
  })

  it('does not render if isShowing set to false', () => {
    const wrapper = mount(KAlert, {
      propsData: {
        'isShowing': false
      }
    })

    expect(wrapper.isEmpty()).toBe(true)
  })
})
