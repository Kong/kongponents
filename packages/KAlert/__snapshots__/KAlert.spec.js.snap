// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KAlert Renders danger variant 1`] = `
<div role="alert" class="k-alert danger" message="I am danger">
  <!---->
  <!----> <span></span></div>
`;

exports[`KAlert Renders info variant 1`] = `
<div role="alert" class="k-alert info" message="I am info">
  <!---->
  <!----> <span></span></div>
`;

exports[`KAlert Renders success variant 1`] = `
<div role="alert" class="k-alert success" message="I am success">
  <!---->
  <!----> <span></span></div>
`;

exports[`KAlert Renders warning variant 1`] = `
<div role="alert" class="k-alert warning" message="I am warning">
  <!---->
  <!----> <span></span></div>
`;
